# Homework 4 - pandas (YOUR NAME HERE)  
### due 3/1/2022   

Use this notebook and prompts to complete the homework. Throughout there will be hints and some code provided  

### Things you will need:  
- Install os, NumPy, pandas  
- states_covid.csv  
- Bloom_etal_2018_Reduced_Dataset.csv  
- logfiles.tgz (or some other multiple file dataset)  

*NOTE*: Make sure your PATH is correct  

**import packages & check required datasets**   

In [109]:
import os
import numpy as np
import pandas as pd

In [110]:
PATH = '/Users/mine/Desktop/Bio792_Shared/program/Part.3.PythonProgramming/Pandas' #CHANGE

In [111]:
cd /Users/mine/Desktop/Bio792_Shared/program/Part.3.PythonProgramming/Pandas/

/Users/mine/Desktop/Bio792_Shared/program/Part.3.PythonProgramming/Pandas


In [112]:
assert os.path.exists(os.path.join(PATH,'states_covid.csv')), 'states_covid.csv does not exist' 

In [113]:
pwd

'/Users/mine/Desktop/Bio792_Shared/program/Part.3.PythonProgramming/Pandas'

In [114]:
assert os.path.exists(os.path.join(PATH,'Bloom_etal_2018_Reduced_Dataset.csv')), 'Bloom_etal_2018_Reduced_Dataset.csv does not exist'

## Task 1 - DataFrame manipulation  

Using **states_covid.csv**, we are going to read the data in as a DataFrame to manipulate, subset, and filter in various ways.   

### Task 1.1 

Read in states_covid.csv with date as a "date" dtype, and only columns consisting of the hospitalization (4 col), ICU (2 col), and Ventilators (2 col)  

In [115]:
states_covid_date_11 = pd.read_csv('states_covid.csv',usecols=['date','hospitalized','hospitalizedCumulative','hospitalizedCurrently','hospitalizedIncrease','inIcuCumulative','inIcuCurrently','onVentilatorCumulative','onVentilatorCurrently'])

In [116]:
states_covid_date_11.dtypes['date']

dtype('O')

### Task 1.2 

For each of the following catergories: *currently* hospitalized, *currently* in the ICU, and *currently* on ventilation...  
Find the 5 states with the greatest numbers in each catergory and list them in order.      
*hint*: sort_values, unique  

In [117]:
state_covid_df = pd.read_csv('states_covid.csv',usecols=['date','hospitalizedCurrently','inIcuCurrently','onVentilatorCurrently','state'],parse_dates=['date'],infer_datetime_format=True)

## make dataframe
states_covid_current = state_covid_df[['date','hospitalizedCurrently','inIcuCurrently','onVentilatorCurrently','state']] ##needs the 2 brackets [[]]. selects these columns only
states_covid_current.head() ## test if works, see columns
#states_covid_current.dtypes

#splitbloom_genus = split_bloom_df.groupby(['genus'],as_index=False).agg('mean') 
    # use groupby
covid_states = states_covid_current.groupby(['state'],as_index=False).mean() ## groups up states
covid_states
covid_states.nlargest(5, ['hospitalizedCurrently','inIcuCurrently','onVentilatorCurrently'])

state  hospitalizedCurrently  inIcuCurrently  onVentilatorCurrently
5     CA            7749.398204     1948.476048                    NaN
47    TX            6251.187879     2274.571429                    NaN
10    FL            4829.707424             NaN                    NaN
37    NY            4506.904070     1107.038806             449.324232
16    IL            2965.941176      687.046440             368.229102

### Task 1.3 

Find the date in which each state crossed 1000 cumulative hospitilized covid patients

In [182]:
date_df = pd.read_csv('states_covid.csv',usecols=['date','hospitalizedCumulative','hospitalizedCurrently','inIcuCurrently','onVentilatorCurrently','state'])
date_df.dtypes
date_df
#date_df.loc

#date_df.loc[date_df['date'] > "1000", 'hospitalizedCumulative']


#date_df.index[date_df['hospitalizedCumulative'].values.argmax()]
date_df.groupby(['state'],as_index=False).mean()
date_df[('hospitalizedCumulative' > 1000).any(1)]




#date_df.groupby(pd.Grouper(key='date')).mean()
#date_df.set_index('date').groupby([pd.Grouper(freq='D'), 'hospitalizedCumulative']).mean()
#df_hospitalcurrent = state_covid_df[['date','hospitalizedCumulative','state']]
#filtered_df_statecovid = df_hospitalcurrent.loc[(df_hospitalcurrent['hospitalizedCumulative'] >=1000)]


TypeError: '>' not supported between instances of 'str' and 'int'

## ignore: trash code that i tried that didn't work

#state_covid_df.hospitalizedCurrently > 1000
#state_covid_hos1000 = state_covid_df[state_covid_df.hospitalizedCurrently > '1000']
#state_covid_hos1000.head()

#NV_covid_df = new_covid_df[new_covid_df.state == 'NV']
#NV_covid_df.head()

#state_covid_hos1000[(state_covid_df.hospitalizedCurrently == 1000)]
#state_covid_df.hospitalizedCurrently > 1000
#NV_covid_df = new_covid_df[new_covid_df.state == 'NV']
#NV_covid_df.head()
#states_covid_current.hospitalizedCurrently > 1000 
#new_covid_df[(new_covid_df.state == 'NV') & (new_covid_df.rate > 0.10)]


## Task 2 - DataFrame summarizing  

Using **Bloom_etal_2018_Reduced_Dataset.csv**, we are going to do more dataframe manipulation and subsetting and summarizing    

### Task 2.1 

Read in Bloom_etal_2018_Reduced_Dataset.csv and create two new columns ('genus','species') that consists of the column *taxa* split at the underscore. Print out the head of this new dataframe and the number of unique genera**     

*hint:* pd.str.split(,expand=True)  

for example:  

| taxa | genus | species   
| :------ | :-- | :---   
| Alosa_alabamae | Alosa | alabamae  

In [12]:
bloom_df = pd.read_csv('Bloom_etal_2018_Reduced_Dataset.csv')
bloom_df.head()

taxa  logbodysize  trophic_position         Reg
0        Alosa_alabamae     1.707570          0.431364  diadromous
1           Alosa_alosa     1.778151          0.556303  diadromous
2          Alosa_fallax     1.778151          0.556303  diadromous
3       Alosa_mediocris     1.778151          0.612784  diadromous
4  Alosa_pseudoharengus     1.602060          0.544068  diadromous

In [88]:
#bloom_df[['genus','species']] = ""# = bloom_df.taxa.str.split("_",expand=True) Makes empty columns
#bloom_df_taxa.head()

df1 = pd.read_csv('Bloom_etal_2018_Reduced_Dataset.csv', usecols=['logbodysize','trophic_position','Reg','taxa'])
df2 = pd.DataFrame(bloom_df.taxa.str.split('_',1).tolist(),columns = ['genus','species']) # new dataframe for making the split
split_bloom_df = pd.concat([df1, df2], axis=1, join='inner') # combining the two dataframes

split_bloom_df.head()

taxa  logbodysize  trophic_position         Reg  genus  \
0        Alosa_alabamae     1.707570          0.431364  diadromous  Alosa   
1           Alosa_alosa     1.778151          0.556303  diadromous  Alosa   
2          Alosa_fallax     1.778151          0.556303  diadromous  Alosa   
3       Alosa_mediocris     1.778151          0.612784  diadromous  Alosa   
4  Alosa_pseudoharengus     1.602060          0.544068  diadromous  Alosa   

          species  
0        alabamae  
1           alosa  
2          fallax  
3       mediocris  
4  pseudoharengus

### Task 2.2 

Create a new dataframe with the mean *logbodysize* and *trophicposition* of each genera. Sort this data frame by the largest body size. Print the head of this dataframe.

In [47]:
## use the split_bloom_df file. Make the index the genus column

splitbloom_genus = split_bloom_df.groupby(['genus'],as_index=False).agg('mean') ### remove the [] around ('mean'). remove genus as the index.
splitbloom_genus.sort_values('logbodysize',ascending=False,inplace=False)

#splitbloom_genus.dtypes

genus  logbodysize  trophic_position
32       Tenualosa     1.778151          0.462398
0            Alosa     1.739062          0.540815
10          Coilia     1.544068          0.477121
15       Ethmalosa     1.544068          0.397940
23      Potamalosa     1.505150          0.518514
17           Hilsa     1.387390          0.462398
5    Anodontostoma     1.342423          0.447158
13       Engraulis     1.190332          0.462398
6     Cetengraulis     1.113943          0.322219
21         Pellona     1.113943          0.623249
3         Anchovia     1.093422          0.477121
33         Thryssa     1.041393          0.447158
22       Pellonula     1.004300          0.504938
14       Escualosa     1.000000          0.505150
16   Gilchristella     1.000000          0.491362
30    Stolothrissa     1.000000          0.431364
18     Hyperlophus     1.000000          0.531479
29     Stolephorus     0.973128          0.518514
7   Chirocentrodon     0.954243          0.518514
12  Encrasicholina     0.903090          0.518514
9       Clupeoides     0.903090          0.491362
24     Ramnogaster     0.903090          0.531479
25   Rhinosardinia     0.903090          0.531479
28   Spratelloides     0.845098          0.491362
2           Anchoa     0.843681          0.522218
19       Jenkinsia     0.838849          0.518514
20    Microthrissa     0.778151          0.491362
4      Anchoviella     0.775207          0.496877
11         Ehirava     0.698970          0.491362
26      Sauvagella     0.672098          0.491362
8     Clupeichthys     0.602060          0.477121
27   Sierrathrissa     0.477121          0.491362
31     Sundasalanx     0.447158          0.491362
1   Amazonsprattus     0.301030          0.531479

Which genera is the smallest and largest?

    Largest: Tenualosa. Smallest: Amazonsprattus

What is the trophic position of the smallest and largest?  

    Largest: 0.462398. Smallest: 0.531479

## Task 3 - Read in muliple files to a dictionary and make a DataFrame - **OPTIONAL/BONUS**  

### This is not something you are expected to do in this course, but just here to give you an idea of the things that you COULD do. Answers will be posted after due date.  


Using **logfiles**: we are going to do read in each file, get some data, append it to a dictionary to later make into a dataframe.     

**note:** *make sure to unzip logfiles*

In [15]:
cd $PATH

/Users/mine/Desktop/Bio792_Shared/program/Part.3.PythonProgramming/Pandas


In [16]:
!tar -xzf logfiles.tgz #unzip logfiles

In [17]:
log_dir = os.path.join(PATH,'logfiles')
assert os.path.exists(log_dir), 'log_dir does not exist'

First step is to find the necessary files. The number of files in the log files is 36, make sure you have that many as well  

In [18]:
!ls -l logfiles/*txt | wc -l 

      36


In [19]:
logfiles = !find $log_dir -name '*txt' #unix command to find files in log_dir directory
logfiles = [os.path.abspath(x) for x in logfiles] #this finds the full path to the file
print(logfiles[1])

/Users/mine/Desktop/Bio792_Shared/program/Part.3.PythonProgramming/Pandas/logfiles/1901302121_H8_S_14.txt.txt


In [20]:
assert(len(logfiles)==36), 'Do not have correct number of logfiles'

In [49]:
cd /Users/mine/Desktop/Bio792_Shared/program/Part.3.PythonProgramming/Pandas/logfiles

/Users/mine/Desktop/Bio792_Shared/program/Part.3.PythonProgramming/Pandas/logfiles


### Getting a little tricky here  

Read in each of the logfiles, for each file extract:  
- minimum temperature  
- maximum temperature  
- date of minimum temp   
- date of maximum temp   
- mean temp for each file.   

This data should all be appended for a dictionary within a for loop:    
Key should be the file name without the path or .txt extension  
Values should be (minTemp,maxTemp,minDate,maxDate,meanTemp)

I recommend making this work for one file first, then putting the rest in a for loop to do the rest.  

Below is an example of how to read in one file

*hint:* do not read date in as date object

In [53]:
### set up data frame as you read in each file
infile = pd.read_csv(logfiles[0],sep='\t',engine='python')
infile.columns = ['Index','Date','Time','Temp','Type']
infile = infile[['Date','Time','Temp']]
infile.head()

Date         Time  Temp
0  9/16/2013   8:00:00 AM  47.9
1  9/16/2013   8:35:00 AM  48.2
2  9/16/2013   9:10:00 AM  48.7
3  9/16/2013   9:45:00 AM  49.4
4  9/16/2013  10:20:00 AM  50.3

In [54]:
infile.dtypes  ## read what type of object date is. How change? what does do not read date as date object mean, change to what?

Date     object
Time     object
Temp    float64
dtype: object

Do everything in steps, make sure it works. Calculate summaries with this one infile:

In [22]:
minTemp =   
maxTemp =  
minDate = infile['Date'][infile['Temp'] == infile['Temp'].min()].unique()[0] #use this for minDate 
maxDate = infile['Date'][infile['Temp'] == infile['Temp'].max()].unique()[0] #use this for maxDate 
meanTemp = 

SyntaxError: invalid syntax (<ipython-input-22-d888355c2242>, line 1)

To get you started, I suggest writing some dummy code in plain words to help outline your for loop:  

In [ ]:
logfiles_dict = {}  
for f in logfiles:  
    #do something   
    #do not read date in as date object  
    #do more something   
    #do other stuff  
    #make print statements EVERYWHERE  
    #append to dict  
    #blahbahblah

In [ ]:
### code here: 

Then do the real code below here. You don't need to turnin your thoughts. Just put it in there as a help reminder. Most people all still do this, no matter how advanced they are

**Once you have created a DataFrame with all the logfiles, print the head and save it to an outfile using pd.to_csv() as logfiles_df.csv** 

#### Here is an example of the final product

In [32]:
logfiles_df.head()

minTemp  maxTemp     minDate    maxDate   meanTemp
1901302235_H6_S_14      21.3     65.8   12/6/2013  7/13/2014  35.726191
1901302121_H8_S_14      17.9     73.4  10/28/2013  7/13/2014  34.698866
1901302217_H18_D_14     21.0     82.4  10/28/2013  7/26/2014  33.334564
1901302110_H13_S_14     22.8     60.9   10/4/2013  7/13/2014  32.240032
1901302146_H14_D_14     24.6     66.1   1/21/2014  7/13/2014  34.242075